onnxruntime/core/optimizer/not_where_fusion.cc

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/optimizer/not_where_fusion.h"
+
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/utils.h"
+
+using namespace ONNX_NAMESPACE;
+using namespace onnxruntime::common;
+namespace onnxruntime {
+
+/**
+Transform that fuses two Not -> Where nodes to a single Where node
+with the where inputs 1 and 2 flipped.
+Condition ->  Not -> Where ->
+              value0-|  |
+              value1----|
+
+Condition -> Where ->
+      value1-|  |
+      value0----|
+ */
+bool NotWhereFusion::SatisfyCondition(const Graph& graph, const Node& node, const logging::Logger& logger) const {
+  if (!graph_utils::IsSupportedOptypeVersionAndDomain(node, "Where", {9})) {
+    return false;
+  }
+
+  const Node* p_not_node = graph_utils::GetInputNode(node, 0);
+  if (p_not_node == nullptr ||
+      !graph_utils::IsSupportedOptypeVersionAndDomain(*p_not_node, "Not", {1}) ||
+      // Make sure the two nodes do not span execution providers.
+      p_not_node->GetExecutionProviderType() != node.GetExecutionProviderType()) {
+    return false;
+  }
+
+  if (p_not_node->GetOutputEdgesCount() > 1) {
+    // all consumers of not must be where
+    for (auto it = p_not_node->OutputNodesBegin(); it != p_not_node->OutputNodesEnd(); ++it) {
+      if (!graph_utils::IsSupportedOptypeVersionAndDomain(*it, "Where", {9})) {
+        return false;
```

Comment:
add one more UT for this ?

Reply:
this single UT graph has all of these cases